### ISSUE TITLE
Zenject and the public npm registry

### ISSUE THREAD

**Issue by kruncher** [ssannandeji/Zenject-2019#262]
I use the npm registry for managing packages and their dependencies in my Unity and web projects.

I understand that you use your own custom solution called Projeny, but it would be really useful if Zenject could also be installed using npm.

I have created a utility that can be used to enable an npm workflow in Unity projects:
https://github.com/rotorz/unity3d-package-syncer
Example package:
https://github.com/rotorz/unity3d-package-example

With the likes of this a brand new Unity project can be created effortlessly with:
```bash
mkdir NewUnityProject
cd NewUnityProject
npm init
******** enter project details ********
******** add "sync": "unity3d--sync" to project's package.json ********
npm install --save unity3d-package-syncer zenject
npm run sync
```

The nice thing with this workflow is that I can install a package via npm that requires zenject; and have zenject automatically get pulled in with it.

After the above install, the project directory structure looks like this:
```
/NewUnityProject/
|-- Assets/
|   |-- Plugins/
|   |   |-- Packages/
|   |   |   |-- zenject/
|   |   |   |   |-- *** Zenject's source files ***
|   |   |   |   |-- package.json
|   |   |   |   |-- README.md
|   |   |   |   |-- LICENSE
|
|-- package.json
|
|-- node_modules/     (excluded in project's .gitignore)
|   |-- a handful of modules supporting 'unity3d-package-syncer'
|   |-- unity3d-package-syncer/
|   |-- zenject/
```

I have created a fork of the Zenject repository and have added a `CreateNpmPackage.bat` file which prepares the directory structure of the npm package:
https://github.com/rotorz/Zenject/blob/master/Build/CreateNpmPackage.bat

Would you be interested in including npm package publishing in your build process so that Zenject can be installed using a simple npm workflow?

Many thanks

**Comment by FullStackForger** [comment]
I am not sure npm is the best manager for c# based projects. NuGet might be better option but after reading http://www.what-could-possibly-go-wrong.com/unity-and-nuget/ I am not sure about it at all.

Why not to simply use asset store?

**Comment by kruncher** [comment]
> I am not sure npm is the best manager for c# based projects. NuGet might be better option but after reading http://www.what-could-possibly-go-wrong.com/unity-and-nuget/ I am not sure about it at all.

We've been using npm for a while and it works really well for us. NuGet works well in the traditional C# development world but Unity projects are more like web-projects in terms of their structure.

For instance, if you run
```sh
mkdir new-project
cd new-project
(echo { "scripts": { "sync": "unity3d--sync" } }) > package.json
npm install --save unity3d-package-syncer @rotorz/unity3d-reorderable-list-examples
npm run sync
```

You'll get everything you need to try out my reorderable list examples in a brand new Unity project:

![image](https://cloud.githubusercontent.com/assets/554215/25920891/49d3bf66-35cb-11e7-8330-d6fb13fe8ee4.png)

**Edit** - I've uploaded a bunch of my modules to github now: https://github.com/rotorz

> Why not to simply use asset store?

- No dependency management
- Slow approval times
- At mercy of asset teams fairly random submission policies
- If a package is updated, redundant asset files are not automatically removed.
- No built-in way to automate pre/post-build tasks.
- Little version management; what if an old project still needs to use an older version of a package...

**Comment by kruncher** [comment]
I'm working on releasing my internal services/configuration library as open source. The only missing piece of the jigsaw puzzle is being able to have Zenject install via `npm`:

WIP: https://github.com/rotorz/unity3d-zenject-services

![](https://cloud.githubusercontent.com/assets/554215/22255889/cf76281c-e250-11e6-9f95-079469881c46.png)

**Comment by kruncher** [comment]
Hmmm, one part of my `CreateNpmPackage.bat` that I would change would be to exclude `OptionalExtras` entirely (edit: done). Instead perhaps publishing those as a separate `zenject-extras` package instead. That way they are opt-in rather than forced-in to every single project.

I just created a brand new project from scratch and realized how chunky that folder is ^_^

**Comment by svermeulen** [comment]
Yep, that's why that folder is there - so that people can easily exclude it.

I think it would be nice to support using Zenject with npm this way.  Can you elaborate more on what that would look like?  Every time we make an update, are you saying that we would also update two git repos? "zenject" and "zenject-extras"?  And each repo would contain a package.json to work properly with npm?

**Comment by kruncher** [comment]
How you structure that would be entirely up to you tbh.

You would only need the one git repo. I think the easiest way to achieve this is to create stubs for the npm packages; and then just copy in the necessary files automatically with a shell script or a bat file or whatever you like to use to manage your build/deploy processes.

At the bare minimum this means having these extra files:

- CreateNpmPackages.bat
- NpmTemplates/zenject/package.json
- NpmTemplates/zenject-extras/package.json

Including any extra npm-specific required files in each of those directories.

All that the `CreateNpmPackages.bat` file has to do is:

1. Create/clean a temporary directory that will hold a copy of the zenject/zenject-extras npm template.
2. Copy the zenject/zenject-extras package template.
3. Replace version number in the `package.json` with latest.
4. Copy relevant files from wherever else in the current git repository.
5. Publish to your npm account on npm.

The npm package is essentially just a "release" of your library. It does not need to include anything else from your git repository. You can specify exclusions in a `.npmignore` file which can exclude things that are only necessary for developing and maintaining your package. An example of this can be things like unit tests that do not need to be included in the release.

It's entirely up to you whether to include the .dll or source versions. I'm personally going with source releases because they'll get bundled into the first-pass.dll anyway due to being inside "Plugins".

The alternative approach would be to structure your git repository in a similar way to mine:
https://github.com/rotorz/unity3d-reorderable-list
https://github.com/rotorz/unity3d-reorderable-list-examples

The advantage with having the repository structured in this way is that during development somebody can easily fork the repository and use the following to install the package from Github (or by specifying the full git url from any) source.

```sh
# The following would install from my Github fork:
npm install --save rotorz/zenject
# The following would install from npm:
npm install --save zenject
# The following would also install from npm:
npm install --save @modesttree/zenject

# In both cases the package would be wired up with the proper `zenject` name
# because `package.json` has a name field which is what gets used regardless
# of where the package came from.
```

In fact, things get even better in the next major release of npm (a few months off I believe) because it can install specific version ranges from git URLs with semver. This will be fantastic for private project-specific (or otherwise internal) packages.

The most important thing is that the package (the .cs editor source files) should not self-modify once installed into a Unity project. Instead generated data files (like configs, etc) should NOT be placed inside the "Assets/Plugins/Packages/{package-name}" directory hierarchy because npm packages are nuked each time they are updated. I am going to be releasing a package soon which makes it easier for packages to create output paths in-editor such as "Assets/Plugins/PackageData/{package-name}/GeneratedConfig.asset".

My tool [unity3d-package-syncer](https://github.com/rotorz/unity3d-package-syncer) detects packages that have the `unity3d-package` keyword and copies the contents of their "assets" folder into the Unity project. This allows the npm package to include non-Unity javascript stuff for automation in addition. For example, in my projects I use `npm run build-iphone` to trigger the iphone build of my project.

I hope that this helps with your question. If not, do not hesitate to throw more questions at me and I'll try my best to assist :)

**Comment by kruncher** [comment]
Using the latest npm it is possible to install my zenject services package using the following (with unity3d-package-syncer setup for the Unity project):
```bash
npm install --save rotorz/unity3d-zenject-services && npm run sync
```

This will install the "rotorz/unity3d-zenject-services" package along with all of it's dependencies (except "zenject") from Github.

![image](https://user-images.githubusercontent.com/554215/31281276-d9010d2e-aaa6-11e7-9b8c-353b46bac944.png)

And then you can setup configuration assets for each of your target platforms using my configuration assets like shown in screenshots in my previous messages.

![image](https://user-images.githubusercontent.com/554215/31281788-5ebf394e-aaa8-11e7-9f7a-d10074e053cd.png)

**Workaround: Since Zenject doesn't support unity3d-package-syncer:**
At the moment the Zenject package has to be faked locally:

1. Create a folder "zenject" (lower case) somewhere on your computer.
2. Add a "package.json" file to the "zenject" folder with.
   ```json
   {
     "name": "zenject",
     "version": "5.2.0",
     "keywords": [ "unity3d-package" ],
     "private": true
   }
   ```
3. Create a sub-folder "assets".
4. Copy "Sources" and "Sources.meta" from zenject into that "assets" folder.
5. With the command line type `npm link` inside the "zenject" folder that was created above.
6. Inside your Unity project package.json manually add the dependency `"zenject": "5.2.0"`.
7. type `npm link zenject` from inside your project directory.

**Comment by kruncher** [comment]
Here are some updates on npm / git package usage.

I have found that the **yarn** package manager (uses the exact same format as npm, and supports all npm and git packages) works a lot better than npm; especially for Unity projects:

1. It doesn't exhibit the regression bug that the npm tool currently still has regarding ssh with git repositories.
2. Packages are installed in exactly the same way each time the yarn tool is used.
3. Packages are installed flat.

I have also found that the **npx** tool that now ships with node/npm simplifies the workflow entirely. It is no longer necessary to add boilerplate to the package.json file in order to sync packages into a Unity project!

For a brand new project:
```bash
yarn init --private
yarn add rotorz/unity3d-tile-system
npx unity3d-package-syncer
```

This will install all dependency packages and sync into the Unity project!!

I will be writing about this to my blog soon, but I thought it would be worth mentioning here :)

In addition I will be adding Unity 2018 .asmdef files to my packages very shortly such that packages can now have the following directory structure:
```
my-package-name/
  |-- assets/
  |    |-- Editor/
  |    |    |-- _assembly.asmdef    (name = my-package-name__editor)
  |    |-- Scripts/
  |    |    |-- _assembly.asmdef    (name = my-package-name)
  |    |-- Tests/
  |    |    |-- _assembly.asmdef    (name = my-package-name__tests)
```

I chose to prefix the .asmdef file with an underscore so that it appears at the top of each script directory; it was totally getting in the way! I chose to use the same name (assembly) for each one since I find this to be a nice and consistent convention.

It would be awesome if Zenject included a .asmdef file with the name "zenject" so that I can have my .asmdef files depend upon it!

**Comment by kruncher** [comment]
Here is a repository that I made which essentially repackages the latest version of Zenject so that it can be installed with `yarn` (or `npm`) and then synced with `unity3d-package-syncer` with support for Unity 2018 assembly definitions:

https://github.com/PoweredBySAM/samlabs-zenject

**Comment by svermeulen** [comment]
Ok I added `zenject.asmdef`, hopefully that helps.  By the way, I also changed to include the zenject-usage.dll directly in git for now to make it easier to for people to sync directly from source

**Comment by kruncher** [comment]
> Ok I added zenject.asmdef, hopefully that helps.

Awesome thanks!

At the moment I am seeing this error:

> Editor/Testing/ZenjectUnitTestFixture.cs(7,7): error CS0246: The type or namespace name `NUnit' could not be found. Are you missing an assembly reference?

> By the way, I also changed to include the zenject-usage.dll directly in git for now to make it easier to for people to sync directly from source

For the source distribution is there any reason not to just have the .cs files inside the "Usage" folder and then just have the .csproj file references them from in there? With the way things are advancing it seems to make sense for packages to be distributed as source rather than compiled assemblies (from within Unity).

Non-Unity projects could then still use the zenject-usage.dll.

**Comment by Zod-** [comment]
@kruncher 
Just a heads up, unity is deploying their own package manager (upm). Right now they only host official packages until the whole process is stable enough and then they'll allow people to add their own packages. As of today It's possible to add a local package like in this [example](https://github.com/Unity-Technologies/ScriptableRenderPipeline/blob/ac1c290b805a4599c0baab5d8cdb1beae36b309c/TestProjects/HDRP_Tests/Packages/manifest.json). It just expects that this folder has a `package.json` with all the needed information. They also said they will allow repository as a package source some time in the future.

More info: https://docs.unity3d.com/Packages/com.unity.package-manager-ui@1.9/manual/index.html

**Comment by kruncher** [comment]
@Zod- Yeah I heard about this, although we have no plans to migrate to their package management offering since yarn/npm does the job very well and are well established and stable tools.
